java/com/google/re2j/Inst.java

```diff
@@ -57,25 +59,37 @@ boolean matchRune(int r) {
     // Special case: single-rune slice is from literal string, not char
     // class.
     if (runes.length == 1) {
-      int r0 = runes[0];
-      if (r == r0) {
-        return true;
-      }
-      if ((arg & RE2.FOLD_CASE) != 0) {
-        for (int r1 = Unicode.simpleFold(r0);
-             r1 != r0;
-             r1 = Unicode.simpleFold(r1)) {
-          if (r == r1) {
-            return true;
-          }
+      return singleMatchRune(r);
+    }
+    
+    return multiMatchRune(r);
+  }
+
+  private boolean singleMatchRune(int r) {
+    int r0 = runes[0];
+    if (r == r0) {
+      return true;
+    }
+
+    if ((arg & RE2.FOLD_CASE) != 0) {
+      int[] folds = Unicode.optimizedFoldOrbit(r0);
+
+      if (folds == null) {
+        return (r == Unicode.optimizedFoldNonOrbit(r0));
```

Comment:
Have you benchmarked this against an implementation based on Character.toLowerCase(r) == r?  Ideally with a few changes like that, we could throw away all the Unicode tables and use the ones built into java.lang.

Be sure to use the int (code point) not char (UTF-16 code) variants of those functions.

Reply:
Yes, the orbit case will still need special handling.  I suggest you do this experiment as a separate change.